src/generator.rs

```diff
@@ -0,0 +1,140 @@
+use std::{
+    fmt,
+    path::{Path, PathBuf},
+};
+
+use eyre::Context;
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    manifest::{Dependency, Manifest},
+    package::PackageStore,
+};
+
+/// The directory used for the generated code
+pub const BUILD_DIRECTORY: &str = "proto/build";
+
+/// The language used for code generation
+#[derive(
+    Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, clap::ValueEnum,
+)]
+#[serde(rename_all = "kebab-case")]
+pub enum Language {
+    Rust,
+}
+
+impl Language {
+    pub fn build_directory(&self) -> PathBuf {
+        Path::new(BUILD_DIRECTORY).join(self.to_string())
+    }
+}
+
+impl fmt::Display for Language {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", serde_typename::to_str(&self).unwrap_or("unknown"))
+    }
+}
+
+/// Backend used to generate code bindings
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Generator {
+    /// The tonic + prost stack
+    Tonic,
+}
+
+impl Generator {
+    pub const TONIC_INCLUDE_FILE: &str = "mod.rs";
+
+    /// Run the generator for a dependency and output files into `out`
+    pub async fn run(&self, dependency: &Dependency, out: &Path) -> eyre::Result<()> {
+        let protoc = protobuf_src::protoc();
+        std::env::set_var("PROTOC", protoc.clone());
+
+        match self {
+            Generator::Tonic => {
+                let out = out.join(dependency.package.as_str());
+
+                fs::remove_dir_all(&out).await.ok();
+
+                fs::create_dir_all(&out)
+                    .await
+                    .wrap_err("failed to recreate dependency output directory")?;
+
+                let package = PackageStore::locate(&dependency.package);
+                let protos = PackageStore::collect(&package).await;
+
+                let includes = &[package];
+
+                tonic_build::configure()
```

Comment:
in this approach, I think you're pulling the tonic code-gen into the buffrs binary. The one concern I'd have here is dependency misalignment between buffrs and tonic - e.g. if if buffrs isn't designed to be run from a build script, what keeps tonic runtime dep + buffrs tonic-build dep in sync? Have you thought about this at all?

Reply:
> buffrs isn't designed to be run from a build script

it is, either cli or `build.rs`

regarding the tonic version you use, we could either reexport tonic for the project, so it only has to use `buffrs::tonic as tonic` to keep them in sync, or trust the user to not mess up there and figure out the compatible versions

Reply:
ack nice, loving the build.rs.

Reply:
What about converting this to a macro so the dependency on `tonic_build` can be removed completely?

Also: building server/client could be made conditional.

Reply:
Configuring buffrs build behavior is meant to go into the manifest `[build.rust]` section!

If this doesnt work for some reason, you can always opt out of the `buffrs::build` plug and play solution